client/my-sites/themes/theme-upload/index.jsx

```diff
@@ -52,26 +119,99 @@ class Upload extends React.Component {
+						<Gridicon
+							className="theme-upload__dropzone-icon"
+							icon="cloud-upload"
+							size={ 48 } />
+						{ dropText }
+					</FilePicker>
+				</div>
+			</div>
+		);
+	}
+
+	renderProgressBar() {
+		const {
+			translate,
+			progressTotal,
+			progressLoaded,
+			installing,
+		} = this.props;
+
+		const uploadingMessage = translate( 'Uploading your theme…' );
+		const installingMessage = translate( 'Installing your theme on site…' );
+
+		return (
+			<div>
+				<span className="theme-upload__title">
+					{ installing ? installingMessage : uploadingMessage }
+				</span>
+				<ProgressBar
+					value={ progressLoaded || 0 }
+					total={ progressTotal || 100 }
+					title={ translate( 'Uploading progress' ) }
+					isPulsing={ installing }
+				/>
+			</div>
+		);
+	}
+
+	renderTheme() {
+		const { uploadedTheme: theme, translate } = this.props;
+		return (
+			<div className="theme-upload__theme-sheet">
+				<span className="theme-upload__theme-name">{ theme.name }</span>
+				<span className="theme-upload__author">
+					{ translate( 'by ' ) }
+					<a href={ theme.author_uri }>{ theme.author }</a>
+				</span>
+				<img src={ theme.screenshot } />
+				<span className="theme-upload__description">{ theme.description }</span>
+			</div>
+		);
+	}
+
+	render() {
+		const { translate, inProgress, complete, failed } = this.props;
+		return (
 			<Main>
 				<HeaderCake onClick={ page.back }>{ translate( 'Upload theme' ) }</HeaderCake>
 				<Card>
-					<span className="theme-upload__prompt">{ uploadPromptText }</span>
-					<span className="theme-upload__instructions">{ uploadInstructionsText }</span>
-					<div className="theme-upload__dropzone">
-						<DropZone onFilesDrop={ this.onFileSelect } />
-						<FilePicker accept="application/zip" onPick={ this.onFileSelect } >
-							<Gridicon
-								className="theme-upload__dropzone-icon"
-								icon="cloud-upload"
-								size={ 48 } />
-							{ dropText }
-						</FilePicker>
-					</div>
+					{ ! inProgress && ! complete && this.renderDropZone() }
+					{ inProgress && this.renderProgressBar() }
+					{ complete && ! failed && this.renderTheme() }
 				</Card>
 			</Main>
 		);
 	}
 }
 
-export default localize( Upload );
-
+export default connect(
+	( state ) => {
+		const siteId = getSelectedSiteId( state );
+		const themeId = getUploadedThemeId( state, siteId );
+		return {
+			siteId,
+			inProgress: isUploadInProgress( state, siteId ),
+			complete: isUploadComplete( state, siteId ),
+			failed: hasUploadFailed( state, siteId ),
+			uploadedTheme: getTheme( state, siteId, themeId ),
+			error: getUploadError( state, siteId ),
+			progressTotal: getUploadProgressTotal( state, siteId ),
+			progressLoaded: getUploadProgressLoaded( state, siteId ),
+			installing: isInstallInProgress( state, siteId ),
+		};
+	},
+	( dispatch ) => {
+		return bindActionCreators( {
+			uploadTheme,
+			clearThemeUpload,
+		}, dispatch );
+	},
```

Comment:
I don't think you need to manually `bindActionCreators` here but can use the `mapDispatchToProps` shorthand:

```js
{ uploadTheme, clearThemeUpload }
```

instead of the entire `( dispatch ) => {...}` statement. This should be functionally equivalent to what you're doing.